feat: support 3rd party dashboard plugins (DHIS2-17283)

edoardo opened the pull request:
Implements [DHIS2-17283](https://dhis2.atlassian.net/browse/DHIS2-17283)

---

### Key features

1. add support for 3rd party dashboard plugins

---

### Description

Dashboard app should have support for custom plugins built with the DHIS2 tooling.
This assumes the plugin uses the platform's components that provide the postRobot communication compatible with this version of dashboard app.

---

### TODO

-   [x] fix Jest tests for AppItem
-   [ ] update min versions for analytics apps

---

### Known issues

-   [ ] tests pass but for AppItem an error happens on render (the updated snapshot makes the test pass)

---

### Screenshots

Example with the Climate data app:
<img width="685" alt="Screenshot 2024-09-06 at 16 21 59" src="https://github.com/user-attachments/assets/8af14659-d7a5-442e-b908-a3d3ae79b9d5">



[DHIS2-17283]: https://dhis2.atlassian.net/browse/DHIS2-17283?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

edoardo committed:
chore: use alpha versions of platform and runtime

edoardo committed:
chore: use alpha versions of platform and runtime

edoardo committed:
refactor: use generic Plugin component from the platform

edoardo committed:
chore: update app-runtime, cli-app-scripts deps

edoardo committed:
chore: update app-runtime, cli-app-scripts deps

edoardo committed:
refactor: handle recording on first load, pass width and height

edoardo committed:
fix: regenerate yarn.lock file

edoardo committed:
fix: regenerate yarn.lock file

edoardo committed:
chore: update app-runtime and cli-app-scripts to latest

edoardo committed:
chore: update app-runtime and cli-app-scripts to latest

edoardo committed:
fix: replace deprecated prop with new one

This silences warnings in the console.

edoardo committed:
fix: avoid warning about nested a tags

Link is rendering an a tag, and MenuItem does the same.
Use a redirect on click instead (already done in the print submenu)

edoardo committed:
fix: silence linter warning

This hopefully has a better solution.

edoardo committed:
refactor: simplify communication for offline caching

It should be enough to just use isParentCached for knowing when to start
recording and removing the cache in the plugins.

edoardo committed:
chore: bump cli-app-scripts dependency

edoardo committed:
chore: bump cli-app-scripts

This should fix some issues with plugin content not being visible.

edoardo committed:
chore: remove unused post-robot dependency

edoardo committed:
chore: fix duplicate app-runtime

edoardo committed:
chore: resolve i18next to <21 to avoid plural strings issues

See https://dhis2.atlassian.net/browse/DHIS2-16591.

edoardo committed:
chore: remove i18next direct dep, resolve i18next-scanner

edoardo committed:
chore: cleanup deps to fix failing build

edoardo committed:
chore: add missing cli-style dep, remove ui resolution

edoardo committed:
fix: fix linting error

edoardo committed:
chore: run prettier to fix linting errors

edoardo committed:
test: update snapshots

edoardo committed:
fix: move visualization fetch in try/catch

This is to avoid the error boundary to show up for uncaught errors
(failed to fetch a visualization) which
interferes with Cypress testing.

edoardo committed:
chore: update cli-app-scripts, remove resolution for i18next-scanner

edoardo committed:
chore: restore lost deps

dependabot[bot] committed:
chore(deps): bump the security group with 3 updates (#3046)

* chore(deps): bump the security group with 3 updates

Updates `ejs` from 3.1.8 to 3.1.10
- [Release notes](https://github.com/mde/ejs/releases)
- [Commits](https://github.com/mde/ejs/compare/v3.1.8...v3.1.10)

Updates `tmpl` from 1.0.4 to 1.0.5
- [Commits](https://github.com/daaku/nodejs-tmpl/commits/v1.0.5)

Updates `word-wrap` from 1.2.3 to 1.2.5
- [Release notes](https://github.com/jonschlinkert/word-wrap/releases)
- [Commits](https://github.com/jonschlinkert/word-wrap/compare/1.2.3...1.2.5)

---
updated-dependencies:
- dependency-name: ejs
  dependency-type: indirect
  dependency-group: security
- dependency-name: tmpl
  dependency-type: indirect
  dependency-group: security
- dependency-name: word-wrap
  dependency-type: indirect
  dependency-group: security
...

Signed-off-by: dependabot[bot] <support@github.com>

* chore: break up command chain

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jen Jones Arnesen <jennifer@dhis2.org>

dependabot[bot] committed:
chore(deps): bump ansi-regex from 3.0.0 to 3.0.1 (#3047)

Bumps [ansi-regex](https://github.com/chalk/ansi-regex) from 3.0.0 to 3.0.1.
- [Release notes](https://github.com/chalk/ansi-regex/releases)
- [Commits](https://github.com/chalk/ansi-regex/compare/v3.0.0...v3.0.1)

---
updated-dependencies:
- dependency-name: ansi-regex
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jen Jones Arnesen <jennifer@dhis2.org>

edoardo committed:
feat: support custom dashboard plugins (DHIS2-17283)

edoardo committed:
chore: fix linting errors

edoardo committed:
chore: restore @dhis2/ui dep that got lost

edoardo committed:
Merge remote-tracking branch 'origin/master' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
Merge remote-tracking branch 'origin/master' into refactor/use-platform-plugin-components

edoardo committed:
Merge branch 'refactor/use-platform-plugin-components' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
chore: remove direct dependency for @dhis2/ui

edoardo committed:
fix: avoid crash if app details are not available

edoardo committed:
refactor: make the missing plugin error more generic

So it can be used also for custom 3rd-party plugins.

edoardo committed:
fix: pass dashboard filters to custom plugins

edoardo committed:
refactor: use same styles for error messages

edoardo committed:
refactor: move item size logic to GridItem

So it can be reused for different item types, ie. both VisualizationItem
and AppItem.

edoardo committed:
refactor: pass width/height to AppItem to fix resize/refresh issue (DHIS2-17947)

The same solution is used in VisualizationItem

edoardo committed:
fix: force fixed dimensions (DHIS2-17947)

This together with the fixes in app-runtime should solve he refresh on
resize issues.

edoardo committed:
refactor: use cached callback to avoid re-render

edoardo committed:
chore: regenerate pot file

edoardo committed:
fix: extend compatible version check to DV/Maps

This is needed as the new plugin system is used, because the messages
used in the communication have changed.

edoardo committed:
chore: regenerate pot file

edoardo committed:
Merge remote-tracking branch 'origin/master' into refactor/use-platform-plugin-components

edoardo committed:
test: fix test for version compatibility

edoardo committed:
chore: regenerate yarn.lock

edoardo committed:
refactor: move error screen to component for easier reuse

edoardo committed:
Merge remote-tracking branch 'origin/refactor/use-platform-plugin-components' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
refactor: move fullscreen functions together in Item folder

edoardo committed:
refactor: move FatalErrorBoundary  and memoizeOne in Item folder

These are now reused for AppItem items too.

edoardo committed:
refactor: adjust imports and use shared fullscreen functions

edoardo committed:
feat: enable View fullscreen and Open in for custom plugin items

edoardo committed:
chore: regenerate pot file

edoardo committed:
refactor: update resize logic

Don't render Plugin before knowing width and height.

edoardo committed:
fix: remove unused prop

Platform's plugin components rely on width and height in order to
disable automatic resize.

edoardo committed:
chore: update deps

edoardo committed:
chore: regenerate pot file

edoardo committed:
test: update snapshots

edoardo committed:
test: update imports and use of data from cached data query

edoardo committed:
chore: fix linting error

edoardo committed:
test: update test for AppItem

Still not quite there.
Currently the render fails with an error, but the updated snapshot makes
the test pass.

jenniferarnesen commented on src/components/Item/ItemMessage/MissingPluginMessage.js:
```diff
@@ -0,0 +1,45 @@
+                <IconWarning24 color={colors.grey500} />
+                <span>
+                    {i18n.t(
+                        'The plugin for rendering this item is not available'
```
Should we use a less technical word than "rendering" ?  Actually, is it even useful to mention that a plugin is missing? Not sure about the second question. If we decide not say "plugin", we could still put that info in a console.warning.

"The plugin needed to display this item is not available."
"This item can't be displayed. Please contact your administrator"

edoardo replied:
I changed the message to your first suggestion.
As for the usefulness of the message, it does make sense for apps that are not bundled (ie. LL) and you open a dashboard which includes a visualization that needs a particular plugin.
In the same message we link to the App Hub and mention the name of the relevant app/plugin.

edoardo committed:
fix: update app-runtime with fix for item resize

edoardo committed:
refactor: set custom title also in edit mode

edoardo committed:
Merge remote-tracking branch 'origin/master' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
fix: convert CSS for rtl support

edoardo committed:
chore: regenerate yarn.lock

edoardo committed:
fix: update next app versions

edoardo committed:
feat: pass dashboardMode to plugins

This is to allow plugins to show different content, ie. an edit button
for configuring the plugin content only when editing a dashboard.
The button won't show when viewing the plugin content.

edoardo committed:
refactor: update message about missing plugin

edoardo committed:
fix: fix URL to App management app

edoardo committed:
Merge remote-tracking branch 'origin/master' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
test: fix app-runtime resolution for Jest, updated snapshot

edoardo committed:
chore: regenerate yarn.lock

edoardo committed:
chore: remove debug log

edoardo committed:
chore: use latest analytics build for testing

edoardo committed:
feat: use delete callback from plugins

This is a promise that the plugin passes to dashboard app.
Regardless of the promise outcome, the dashboard item is always deleted,
but only after calling the promise.
This gives plugins a chance to do some cleanup before the iframe is
removed from the DOM.

edoardo committed:
test: fix failing test

edoardo committed:
fix: fix bug which caused 3rd party plugins to not always load

edoardo committed:
Merge remote-tracking branch 'origin/master' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
refactor: wrap ref callback in useCallback

edoardo committed:
refactor: fix sonarcloud issues

edoardo committed:
fix: fix arguments order

This is related to the previous fix of the default initialisation of
appDetails.

edoardo committed:
test: fix tests after refactoring

edoardo committed:
refactor: remove unnecessary filterVersion prop

edoardo committed:
fix: make filterVersion actually change

edoardo committed:
Merge branch 'master' into feat/support-custom-plugin-DHIS2-17283

edoardo committed:
refactor: use same style for errors

Remove the Cover component as it does not cover anything...

edoardo committed:
test: fix test snapshot

edoardo committed:
chore: remove unused code

edoardo committed:
chore: remove unused import

edoardo committed:
chore: update ui dependency

jenniferarnesen commented:
Superseded by https://github.com/dhis2/dashboard-app/pull/3184
